File: apps/ios/Snout/Utilities/Money.swift
//
//  Money.swift
//  Snout
//
//  Web parity: src/lib/money.ts
//  Both implementations must satisfy the same inputs and outputs.
//
//  Cents-based math + currency formatting. Currency is per-organization (set at signup).
//  Don't ask the user to choose currency at the iOS layer.
//

import Foundation

enum Money {
    /// `formatCents(cents, "CAD")` → "$52.50 CAD"
    /// Mirrors the web app's deliberately simple format. Not using NumberFormatter.currency
    /// because the web app's representation is "$<value> <code>" and parity is the goal.
    static func formatCents(_ cents: Int?, currency: String = "CAD") -> String {
        let value = Double(cents ?? 0) / 100.0
        return String(format: "$%.2f %@", value, currency)
    }

    /// `formatCentsShort(cents)` → "$52.50"
    static func formatCentsShort(_ cents: Int?) -> String {
        let value = Double(cents ?? 0) / 100.0
        return String(format: "$%.2f", value)
    }

    /// Parse a user-entered dollar amount like "52.50" into integer cents (5250).
    /// Returns nil on invalid input.
    static func parseDollarsToCents(_ input: String) -> Int? {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return nil }
        guard let n = Double(trimmed), n.isFinite, n >= 0 else { return nil }
        return Int((n * 100).rounded())
    }

    /// `centsToDollarString(cents)` → "52.50" (or "" when cents is nil)
    static func centsToDollarString(_ cents: Int?) -> String {
        guard let cents else { return "" }
        let value = Double(cents) / 100.0
        return String(format: "%.2f", value)
    }

    static func formatDurationType(_ t: String) -> String {
        switch t {
        case "hourly":      return "Hourly"
        case "half_day":    return "Half Day"
        case "full_day":    return "Full Day"
        case "overnight":   return "Overnight"
        case "multi_night": return "Multi-Night"
        default:            return t
        }
    }

    static func formatModule(_ m: String) -> String {
        switch m {
        case "daycare":  return "Daycare"
        case "boarding": return "Boarding"
        case "grooming": return "Grooming"
        case "training": return "Training"
        case "retail":   return "Retail"
        default:         return m
        }
    }

    static func formatReservationStatus(_ s: String) -> String {
        switch s {
        case "requested":    return "Requested"
        case "confirmed":    return "Confirmed"
        case "checked_in":   return "Checked In"
        case "checked_out":  return "Checked Out"
        case "cancelled":    return "Cancelled"
        case "no_show":      return "No Show"
        default:             return s
        }
    }
}

File: apps/web/src/lib/money.ts
export function formatCents(cents: number | null | undefined, currency: string = "CAD") {
  const v = (Number(cents ?? 0) / 100).toFixed(2);
  return `$${v} ${currency}`;
}

export function formatCentsShort(cents: number | null | undefined) {
  const v = (Number(cents ?? 0) / 100).toFixed(2);
  return `$${v}`;
}

/** Parse a user-entered dollar amount like "52.50" into integer cents (5250). Returns null on invalid. */
export function parseDollarsToCents(input: string): number | null {
  const trimmed = input.trim();
  if (!trimmed) return null;
  const n = Number(trimmed);
  if (!Number.isFinite(n) || n < 0) return null;
  return Math.round(n * 100);
}

export function centsToDollarString(cents: number | null | undefined): string {
  if (cents == null) return "";
  return (Number(cents) / 100).toFixed(2);
}

export function formatDurationType(t: string): string {
  const map: Record<string, string> = {
    hourly: "Hourly",
    half_day: "Half Day",
    full_day: "Full Day",
    overnight: "Overnight",
    multi_night: "Multi-Night",
  };
  return map[t] ?? t;
}

export function formatModule(m: string): string {
  const map: Record<string, string> = {
    daycare: "Daycare",
    boarding: "Boarding",
    grooming: "Grooming",
    training: "Training",
    retail: "Retail",
  };
  return map[m] ?? m;
}

export function formatReservationStatus(s: string): string {
  const map: Record<string, string> = {
    requested: "Requested",
    confirmed: "Confirmed",
    checked_in: "Checked In",
    checked_out: "Checked Out",
    cancelled: "Cancelled",
    no_show: "No Show",
  };
  return map[s] ?? s;
}

export function formatDateTime(value: string | Date | null | undefined, timezone?: string) {
  if (!value) return "—";
  const d = typeof value === "string" ? new Date(value) : value;
  return d.toLocaleString(undefined, {
    timeZone: timezone || undefined,
    year: "numeric",
    month: "short",
    day: "numeric",
    hour: "numeric",
    minute: "2-digit",
  });
}

export function formatTime(value: string | Date | null | undefined, timezone?: string) {
  if (!value) return "—";
  const d = typeof value === "string" ? new Date(value) : value;
  return d.toLocaleTimeString(undefined, {
    timeZone: timezone || undefined,
    hour: "numeric",
    minute: "2-digit",
  });
}

/** Build a default end_at from a start_at string (datetime-local) and a duration_type. */
export function computeEndFromStart(startISO: string, durationType: string): string {
  const start = new Date(startISO);
  if (Number.isNaN(start.getTime())) return startISO;
  const end = new Date(start);
  switch (durationType) {
    case "hourly":
      end.setHours(end.getHours() + 1);
      break;
    case "half_day":
      end.setHours(end.getHours() + 5);
      break;
    case "full_day":
      end.setHours(end.getHours() + 10);
      break;
    case "overnight":
      end.setDate(end.getDate() + 1);
      end.setHours(8, 0, 0, 0);
      break;
    case "multi_night":
      end.setDate(end.getDate() + 1);
      break;
    default:
      end.setHours(end.getHours() + 1);
  }
  // Format as datetime-local string
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${end.getFullYear()}-${pad(end.getMonth() + 1)}-${pad(end.getDate())}T${pad(end.getHours())}:${pad(end.getMinutes())}`;
}

/** Convert a Date or ISO to a value usable in <input type="datetime-local"> */
export function toDatetimeLocalValue(value: string | Date | null | undefined): string {
  if (!value) return "";
  const d = typeof value === "string" ? new Date(value) : value;
  if (Number.isNaN(d.getTime())) return "";
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}T${pad(d.getHours())}:${pad(d.getMinutes())}`;
}
